apps/hash-ai-worker-py/app/typesystem/_schema.py

```diff
@@ -14,21 +33,118 @@ class OntologyTypeSchema(BaseModel, ABC):
     schema_url: str = Field(..., alias="$schema")
 
 
-T = TypeVar("T", bound=BaseModel)
+T = TypeVar("T", bound=Schema)
 
 
-class OneOf(BaseModel, Generic[T]):
+class OneOf(Schema, Generic[T]):
     one_of: list[T] = Field(..., alias="oneOf")
 
+    async def create_model(
+        self,
+        *,
+        actor_id: UUID,
+    ) -> Annotated[Any, ...]:
+        types = await asyncio.gather(
+            *[value.create_model(actor_id=actor_id) for value in self.one_of],
+        )
+
+        class OneOfSchema:
+            @classmethod
+            def __get_pydantic_core_schema__(
+                cls,
+                _source_type: Any,  # noqa: ANN401
+                handler: GetCoreSchemaHandler,
+            ) -> CoreSchema:
+                return core_schema.union_schema([handler(t) for t in types])
+
+            @classmethod
+            def __get_pydantic_json_schema__(
+                cls,
+                schema: CoreSchema,
+                handler: GetJsonSchemaHandler,
+            ) -> JsonSchemaValue:
+                json_schema = handler(schema)
+                if any_of := json_schema.pop("anyOf", None):
+                    json_schema["oneOf"] = any_of
+                return json_schema
 
-class Array(BaseModel, Generic[T]):
+        return Annotated[OneOf[T], OneOfSchema]
+
+
+class Array(Schema, Generic[T]):
     ty: Literal["array"] = Field(..., alias="type")
     items: T
     min_items: int | None = Field(default=None, alias="minItems")
     max_items: int | None = Field(default=None, alias="maxItems")
 
+    async def create_model(
+        self,
+        *,
+        actor_id: UUID,
+    ) -> Annotated[Any, ...]:
+        ty = await self.items.create_model(actor_id=actor_id)
+
+        class ListSchema:
+            @classmethod
+            def __get_pydantic_core_schema__(
+                cls,
+                _source_type: Any,  # noqa: ANN401
+                handler: GetCoreSchemaHandler,
+            ) -> CoreSchema:
+                return core_schema.list_schema(
+                    handler.generate_schema(ty),
+                    min_length=self.min_items,
+                    max_length=self.max_items,
+                )
 
-class Object(BaseModel, Generic[T]):
+        return Annotated[list[T], ListSchema]
+
+
+class Object(Schema, Generic[T]):
     ty: Literal["object"] = Field(..., alias="type")
     properties: dict[str, T]
     required: list[str] | None = None
+
+    async def create_model(
+        self,
+        *,
+        actor_id: UUID,
+    ) -> Annotated[Any, ...]:
+        types = {
+            key: await value.create_model(actor_id=actor_id)
```

Comment:
This should also make use of `asyncio.gather`

Reply:
How does `gather` work with dictionaries?

Reply:
There is no easy way, sadly; you would need to first gather all key,value and then construct the dict from that.

I am calling this out as this will make HTTP calls, which should be parallelized.